### PR TITLE
ports/mimxrt: Add support for UART hardware flow control.

### PR DESCRIPTION
**Pull request by iabdalkader**

### CODE REVIEW

**Comment by robert-hh** [comment]
That addition looks familiar A year ago I had already coded flow control for the MIMXRT boards. I did not make a PR at that time, after I noticed that the only boards which had the RTS/CTS pins at least for one UART available at board level were the Teesies. I cannot even recall whether it were both. But it may help for a future board design. Are you doing that ATM?

**Comment by iabdalkader** [comment]
> But it may help for a future board design. Are you doing that ATM?

Do you mean if we use it ? Yes our new board uses both RTS/CTS for CYW43 bluetooth, to support the highest baudrate.

**Comment by robert-hh** [comment]
> Yes our new board uses both RTS/CTS for CYW43 bluetooth

That's what it sounded like. And there you got in touch with the Pin handling, which is awkward, especially how modules like UART, SPI, I2C are defined in mpconfigboard.h

**Comment by iabdalkader** [comment]
> And there you got in touch with the Pin handling, which is awkward, especially how modules like UART, SPI, I2C are defined in mpconfigboard.h

For now we followed what was already in the port (tables) but I still have a better proposal for pin handling, just don't have the time to send it to micropython and follow-up on it.

File: ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h [hunk 1]
```diff
@@ -23,6 +23,12 @@
     { IOMUXC_GPIO_08_LPUART3_TXD }, { IOMUXC_GPIO_AD_07_LPUART3_RXD }, \
     { IOMUXC_GPIO_AD_02_LPUART4_TXD }, { IOMUXC_GPIO_AD_01_LPUART4_RXD },
 
+#define IOMUX_TABLE_UART_CTS_RTS \
+    { IOMUXC_GPIO_08_LPUART1_CTS_B }, { IOMUXC_GPIO_07_LPUART1_RTS_B }, \
+    { 0 }, { 0 }, \
+    { 0 }, { 0 }, \
```
Comment (dpgeorge) [inline ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h:29]: What happens if this third uart is used with flow control enabled, will it detect that flow is not supported on this uart and raise a Python exception?

File: ports/mimxrt/machine_uart.c [hunk 2]
```diff
@@ -98,6 +108,32 @@ bool lpuart_set_iomux(int8_t uart) {
     }
 }
 
+bool lpuart_set_iomux_rts(int8_t uart) {
+    int index = (uart - 1) * 2;
```
Comment (dpgeorge) [inline ports/mimxrt/machine_uart.c:112]: Might be a good idea to put a static assertion in this function to assert that the size of `iomux_table_uart` and `ioumux_table_uart_cts_rts` tables are the same.

File: ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h (hunk 1 above)
Comment (iabdalkader) [reply ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h:29]: Yes it should raise an exception if RTS or CTS mux register is zero.

File: ports/mimxrt/machine_uart.c (hunk 2 above)
Comment (iabdalkader) [reply ports/mimxrt/machine_uart.c:113]: Done, also fixed a typo (/cst/cts/)

**Comment by dpgeorge** [comment]
@iabdalkader it looks like this broke two boards:
```
CC machine_uart.c
In file included from ./mpconfigport.h:30,
                 from ../../py/mpconfig.h:75,
                 from ../../py/mpstate.h:31,
                 from ../../py/runtime.h:29,
                 from machine_uart.c:28:
boards/OLIMEX_RT1010/mpconfigboard.h:29:7: error: 'IOMUXC_GPIO_AD_B0_14_LPUART1_CTS_B' undeclared here (not in a function); did you mean 'IOMUXC_GPIO_AD_14_LPUART4_CTS_B'?
   29 |     { IOMUXC_GPIO_AD_B0_14_LPUART1_CTS_B }, { IOMUXC_GPIO_AD_B0_15_LPUART1_RTS_B }, \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
machine_uart.c:82:5: note: in expansion of macro 'IOMUX_TABLE_UART_CTS_RTS'
   82 |     IOMUX_TABLE_UART_CTS_RTS
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
boards/OLIMEX_RT1010/mpconfigboard.h:29:47: error: 'IOMUXC_GPIO_AD_B0_15_LPUART1_RTS_B' undeclared here (not in a function); did you mean 'IOMUXC_GPIO_AD_07_LPUART2_RTS_B'?
   29 |     { IOMUXC_GPIO_AD_B0_14_LPUART1_CTS_B }, { IOMUXC_GPIO_AD_B0_15_LPUART1_RTS_B }, \
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
machine_uart.c:82:5: note: in expansion of macro 'IOMUX_TABLE_UART_CTS_RTS'
   82 |     IOMUX_TABLE_UART_CTS_RTS
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
boards/OLIMEX_RT1010/mpconfigboard.h:32:7: error: 'IOMUXC_GPIO_EMC_17_LPUART4_CTS_B' undeclared here (not in a function); did you mean 'IOMUXC_GPIO_AD_14_LPUART4_CTS_B'?
   32 |     { IOMUXC_GPIO_EMC_17_LPUART4_CTS_B }, { IOMUXC_GPIO_EMC_18_LPUART4_RTS_B },
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
machine_uart.c:82:5: note: in expansion of macro 'IOMUX_TABLE_UART_CTS_RTS'
   82 |     IOMUX_TABLE_UART_CTS_RTS
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
boards/OLIMEX_RT1010/mpconfigboard.h:32:45: error: 'IOMUXC_GPIO_EMC_18_LPUART4_RTS_B' undeclared here (not in a function); did you mean 'IOMUXC_GPIO_AD_13_LPUART4_RTS_B'?
   32 |     { IOMUXC_GPIO_EMC_17_LPUART4_CTS_B }, { IOMUXC_GPIO_EMC_18_LPUART4_RTS_B },
      |                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
machine_uart.c:82:5: note: in expansion of macro 'IOMUX_TABLE_UART_CTS_RTS'
   82 |     IOMUX_TABLE_UART_CTS_RTS
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../py/mpstate.h:33:
machine_uart.c: In function 'lpuart_set_iomux_rts':
../../py/misc.h:54:50: error: size of unnamed array is negative
   54 | #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
      |                                                  ^
machine_uart.c:112:5: note: in expansion of macro 'MP_STATIC_ASSERT'
  112 |     MP_STATIC_ASSERT(MP_ARRAY_SIZE(iomux_table_uart) == MP_ARRAY_SIZE(iomux_table_uart_cts_rts));
      |     ^~~~~~~~~~~~~~~~
See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../../py/mkrules.mk:83: build-OLIMEX_RT1010/machine_uart.o] Error 1
```
and
```
machine_uart.c:82:5: error: 'IOMUX_TABLE_UART_CTS_RTS' undeclared here (not in a function); did you mean 'IOMUX_TABLE_UART'?
   82 |     IOMUX_TABLE_UART_CTS_RTS
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
      |     IOMUX_TABLE_UART
In file included from ../../py/mpstate.h:33,
                 from ../../py/runtime.h:29,
                 from machine_uart.c:28:
machine_uart.c: In function 'lpuart_set_iomux_rts':
../../py/misc.h:54:50: error: size of unnamed array is negative
   54 | #define MP_STATIC_ASSERT(cond) ((void)sizeof(char[1 - 2 * !(cond)]))
      |                                                  ^
machine_uart.c:112:5: note: in expansion of macro 'MP_STATIC_ASSERT'
  112 |     MP_STATIC_ASSERT(MP_ARRAY_SIZE(iomux_table_uart) == MP_ARRAY_SIZE(iomux_table_uart_cts_rts));
      |     ^~~~~~~~~~~~~~~~
See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
make: *** [../../py/mkrules.mk:83: build-ADAFRUIT_METRO_M7/machine_uart.o] Error 1
```
Are you able to fix them?

**Comment by robert-hh** [comment]
Sorry, I did not cross-check all boards, just all MCUs. These two shoudl be the same as the MIMXRT1010_EVK. @iabdalkader, if you do not have time/access to the files ATM, I can fix it.

**Comment by iabdalkader** [comment]
@robert-hh @dpgeorge Hi, sorry about that, yes we can fix it, it could take me a few days to get around to it, I can mention @ kwagyeman to fix it (he did the work) or @robert-hh if you'd like, and have the time, to fix it please go ahead.

**Comment by robert-hh** [comment]
@iabdalkader I made already a PR with the fix, knowing that you are on vacation.

**Pull request merged**